File: ai-service/app/schemas/github.py
"""Pydantic schemas for the GitHub onboarding scan (roles/01, 01a).

Two groups:
  1. Public result models returned to the TS backend (mirror the DynamoDB
     tables: freelancer_skills, freelancer_projects, profile_confidence,
     github_scan_jobs).
  2. Small LLM-only schemas used for the last-mile semantic step (skill
     name normalization + project summaries). Facts never go through the LLM.
"""
from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, Field

SkillCategory = Literal["language", "framework", "tool", "domain"]
ConfidenceBand = Literal["emerging", "developing", "match_ready"]
SegmentState = Literal["pending", "running", "done", "fallback", "error"]


# ───────────────────────── Request ─────────────────────────

class GithubScanRequest(BaseModel):
    githubUsername: str = Field(min_length=1)
    # The freelancer's OAuth access token (from GitHub sign-in). GraphQL needs a
    # token; if omitted, the service falls back to the server GITHUB_TOKEN.
    accessToken: Optional[str] = None
    topN: Optional[int] = Field(default=None, ge=1, le=200)


# ───────────────────────── Segment outputs ─────────────────────────

class SkillEvidence(BaseModel):
    repo: str
    signal: str          # e.g. "language", "dependency", "topic"
    detail: str          # e.g. "812kb TypeScript", "depends on: react"


class VerifiedSkill(BaseModel):
    name: str
    category: SkillCategory
    confidence: int = Field(ge=0, le=100)
    evidence: List[SkillEvidence] = Field(default_factory=list)
    source: Literal["github_scan"] = "github_scan"
    editable: Literal[False] = False   # tamper-proof by construction


class FreelancerProject(BaseModel):
    repoName: str
    summary: str
    domain: Optional[str] = None
    stack: List[str] = Field(default_factory=list)
    stars: int = 0
    commitShare: int = 0               # 0-100 (ownership/authorship proxy)
    lastActiveAt: Optional[str] = None
    rankScore: int = 0


class ExperienceSignals(BaseModel):
    totalCommits: int = 0              # commits AUTHORED by the user (measured)
    reposAnalyzed: int = 0
    activeYears: float = 0.0
    avgStars: float = 0.0
    collaborationRepos: int = 0        # others' repos where the user authored commits
    documentationQuality: int = Field(default=0, ge=0, le=100)
    linesAuthored: int = 0             # net lines the user wrote (top repos, best-effort)
    pullRequests: int = 0              # PRs opened in the trailing year
    accountAgeYears: float = 0.0       # GitHub account tenure
    followers: int = 0


class ConfidenceFactorBreakdown(BaseModel):
    skillBreadthDepth: int = 0
    projectStrength: int = 0
    recency: int = 0
    contributionVolume: int = 0
    documentation: int = 0


class ProfileConfidence(BaseModel):
    score: int = Field(ge=0, le=100)
    band: ConfidenceBand
    factorBreakdown: ConfidenceFactorBreakdown


class SegmentStatus(BaseModel):
    skills: SegmentState = "pending"
    projects: SegmentState = "pending"
    experience: SegmentState = "pending"


# ───────────────────────── Full result ─────────────────────────

class GithubScanResult(BaseModel):
    githubUsername: str
    reposDiscovered: int
    reposAnalyzed: int
    languages: dict[str, int] = Field(default_factory=dict)  # name -> percent
    skills: List[VerifiedSkill] = Field(default_factory=list)
    projects: List[FreelancerProject] = Field(default_factory=list)
    experience: ExperienceSignals = Field(default_factory=ExperienceSignals)
    confidence: ProfileConfidence
    segmentStatus: SegmentStatus
    scannedAt: str


# ───────────────────────── LLM-only schemas (last-mile) ─────────────────────────

class NormalizedSkill(BaseModel):
    raw: str
    normalized: str
    category: SkillCategory


class SkillNormalizationOutput(BaseModel):
    """LLM cleans/dedupes long-tail skill tokens the lookup table missed."""
    skills: List[NormalizedSkill]


class ProjectSummary(BaseModel):
    repoName: str
    summary: str = Field(min_length=1)
    domain: str


class ProjectSummariesOutput(BaseModel):
    """One batched call summarizes all top projects at once."""
    projects: List[ProjectSummary]

File: backend/src/types/github.ts
/**
 * TypeScript mirrors of the Python `ai-service` GitHub-scan schemas
 * (ai-service/app/schemas/github.py). Kept in sync field-for-field so the
 * gateway can persist and re-emit scan results without re-deriving anything.
 */

export interface SkillEvidence {
  repo: string;
  signal: string;
  detail: string;
}

export interface VerifiedSkill {
  name: string;
  category: 'language' | 'framework' | 'tool' | 'domain';
  confidence: number;
  evidence: SkillEvidence[];
  source: 'github_scan';
  editable: false;
}

export interface FreelancerProject {
  repoName: string;
  summary: string;
  domain?: string | null;
  stack: string[];
  stars: number;
  commitShare: number;
  lastActiveAt?: string | null;
  rankScore: number;
}

export interface ExperienceSignals {
  totalCommits: number;          // commits AUTHORED by the user (measured, not all authors)
  reposAnalyzed: number;
  activeYears: number;
  avgStars: number;
  collaborationRepos: number;    // others' repos where the user authored commits
  documentationQuality: number;
  linesAuthored?: number;        // net lines the user wrote (top repos, best-effort)
  pullRequests?: number;         // PRs opened in the trailing year
  accountAgeYears?: number;      // GitHub account tenure
  followers?: number;
}

export interface ConfidenceFactorBreakdown {
  skillBreadthDepth: number;
  projectStrength: number;
  recency: number;
  contributionVolume: number;
  documentation: number;
}

export interface ProfileConfidence {
  score: number;
  band: 'emerging' | 'developing' | 'match_ready';
  factorBreakdown: ConfidenceFactorBreakdown;
}

export type SegmentState = 'pending' | 'running' | 'done' | 'fallback' | 'error';

export interface SegmentStatus {
  skills: SegmentState;
  projects: SegmentState;
  experience: SegmentState;
}

/** Full result returned by the AI service `/ai/github/scan`. */
export interface GithubScanResult {
  githubUsername: string;
  reposDiscovered: number;
  reposAnalyzed: number;
  languages: Record<string, number>;
  skills: VerifiedSkill[];
  projects: FreelancerProject[];
  experience: ExperienceSignals;
  confidence: ProfileConfidence;
  segmentStatus: SegmentStatus;
  scannedAt: string;
}

export type ScanJobStatus = 'queued' | 'running' | 'partial' | 'complete' | 'failed';

/** Persisted scan-job record (github_scan_jobs table). */
export interface ScanJob {
  jobId: string;
  freelancerId: string;
  githubUsername: string;
  status: ScanJobStatus;
  segmentStatus: SegmentStatus;
  reposDiscovered: number;
  reposAnalyzed: number;
  languages?: Record<string, number>;
  experience?: ExperienceSignals | null;
  confidence?: ProfileConfidence | null;
  error?: string | null;
  createdAt: string;
  updatedAt: string;
  finishedAt?: string | null;
}

/** A single SSE frame from the scan stream. */
export interface ScanStreamEvent {
  event: 'scan_started' | 'segment_ready' | 'scan_complete' | 'scan_error';
  data: any;
}
